### PR TITLE
feat: add token caching

### PR DESCRIPTION
**Pull request by aymeric-dispa**
The purpose of this PR is to add token caching for the GO SDK.
With this cache, newly created connections will reused an existing access token (when available), and fetch a new one if the token is not available. Also, a new token will be fetched upon expiry or when a 401 is returned by the server.

client_test.go was added in order to test the lifecycle of the tokens (when fetched, when they expired and when the server returns a 401)

### CODE REVIEW

File: connection_integration_test.go [hunk 1]
```diff
@@ -22,11 +22,6 @@ func TestConnectionSetStatement(t *testing.T) {
 	_, err = conn.QueryContext(context.TODO(), "SELECT * FROM information_schema.tables", nil)
 	assert(err == nil, t, "query returned an error, but shouldn't")
 
-	_, err = conn.ExecContext(context.TODO(), "SET use_standard_sql=0", nil)
-	assert(err == nil, t, "set use_standard_sql returned an error, but shouldn't")
-
-	_, err = conn.QueryContext(context.TODO(), "SELECT * FROM information_schema.tables", nil)
-	assert(err != nil, t, "query didn't return an error, but should")
```
Comment (ptiurin) [inline connection_integration_test.go:29]: Do those not work anymore?

File: client.go [hunk 2]
```diff
@@ -205,11 +206,35 @@ func checkErrorResponse(response []byte) error {
...
+	accessToken := getCachedAccessToken(c.Username, c.ApiEndpoint)
+	var response []byte
+	var err error
+	var responseCode int
+	if len(accessToken) > 0 {
+		response, err, responseCode = request(ctx, accessToken, method, url, userAgent, params, bodyStr)
+	}
+	if len(accessToken) == 0 || responseCode == http.StatusUnauthorized {
+		deleteAccessTokenFromCache(c.Username, c.ApiEndpoint)
+		// Refreshing the access token as it is expired
```
Comment (ptiurin) [inline client.go:221]: Can we check for expired token before making a request? Since we already have an expiration time cached it would be more efficient than doing a round trip to the server.

**Review by ptiurin — COMMENTED** [review]
Just some comments and clarifications.
Also, what does this change solve? Do we have a use case where multiple connections are established in the same session?
In Python SDK we've added an on-disk token cache that would be read into memory on connection initialisation. This allows running subsequent programs without authenticating each time. Should we do something similar for the GO SDK as well?

File: client.go (hunk 2 above)
Comment (aymeric-dispa) [reply client.go:221]: getCachedAccessToken(c.Username, c.ApiEndpoint) returns tokens that are not expired based on their expiration time (ttlcache gets rid of the tokens that are expired automatically) . We could face a  'http.StatusUnauthorized' if the token is 'rejected' by the server although based on their expiration time they should still be valid(which happened once in the past).

**Comment by aymeric-dispa** [comment]
> Just some comments and clarifications. Also, what does this change solve? Do we have a use case where multiple connections are established in the same session? In Python SDK we've added an on-disk token cache that would be read into memory on connection initialisation. This allows running subsequent programs without authenticating each time. Should we do something similar for the GO SDK as well?

It solves the issue where multiple connections are needed when using multithreading. Without the cache we hit the rate limit very quickly (I actually faced the issue when trying to implement performance tests). 

Caching in a local file is a lot less performant and having the token cached in memory simplify the implementation (no security issue of having a token in a file, not having to deal with the lifecycle of the token when it is expired as it's handled by the lib and no potential permission issue when trying to write a file in the disk of the host)

File: connection_integration_test.go (hunk 1 above)
Comment (aymeric-dispa) [reply connection_integration_test.go:29]: No 😞 
image

File: auth.go [hunk 3]
```diff
@@ -13,27 +16,54 @@ type AuthenticationResponse struct {
...
-		return nil, ConstructNestedError("authentication request failed", err)
+		infolog.Printf("Authentication was successful")
+		cache.Set(getCacheKey(username, apiEndpoint), authResp.AccessToken, time.Duration(authResp.ExpiresIn)*time.Millisecond)
+		return &Client{Username: username, Password: password, ApiEndpoint: apiEndpoint, UserAgent: userAgent}, nil
 	}
+}
+
+// getCacheKey calculates an access token key using the username and the apiEndpoint provided
+func getCacheKey(username, apiEndpoint string) string {
+	return username + apiEndpoint
```
[the remark this answers is not in the document]
Comment (aymeric-dispa) [reply auth.go:53]: We could hash but I'm not sure it would bring benefits (all the data is available in the dsn so I don't think there is a point hiding the data in memory). Also, it makes things easier when trying to debug (eg, if a test fails we can put a breaking and check the content of the cache if needed)

File: client.go (hunk 2 above)
Comment (ptiurin) [reply client.go:221]: I see. Looking through ttlcache readme I think we're supposed to call `cache.Start()` or `cache.DeleteExpired()` to evict entries? Or do you mean `getCachedAccessToken` would not return invalid entry and we'll just overwrite it with a new token, instead of deleting it?

**Comment by ptiurin** [comment]
> It solves the issue where multiple connections are opened (eg: when using multithreading). Without the cache we hit the rate limit very quickly (I actually faced the issue when trying to implement performance tests).

Will this be the case in a real-world scenario as well? Generally a connection is shared across threads in a multi-threaded environment to avoid the overhead of setting it up as well as re-auth. Is this something Go does not support?

> Caching in a local file is less performant and having the token cached in memory simplifies the implementation (no security issue of having a token in a file, no having to deal with the lifecycle of the token when it is expired as it's handled by the lib and no potential permission issue when trying to write a file in the disk of the host).

I see your point, but these are the issues that we've already faced with the Python SDK. Reading from file is slow, but the price is paid only on startup or when token is changed, otherwise it's cached in memory. In there token caching operates on a best effort scenario, if the disk is read-only it does not fail and instead caches in memory.
The bottom line is, I'd rather have a consistent approach to caching across different SDKs than each one doing its own thing.

File: client.go (hunk 2 above)
Comment (aymeric-dispa) [reply client.go:221]: cache.Start() and cache.DeleteExpired() are not needed when we set the ttl during the insertion of the token in the cache. (This is tested by https://github.com/firebolt-db/firebolt-go-sdk/blob/6749d13ec592b525eca60ee855a180e54c71bad8/client_test.go#L83-L116 ).

**Comment by aymeric-dispa** [comment]
> > It solves the issue where multiple connections are opened (eg: when using multithreading). Without the cache we hit the rate limit very quickly (I actually faced the issue when trying to implement performance tests).
> 
> Will this be the case in a real-world scenario as well? Generally a connection is shared across threads in a multi-threaded environment to avoid the overhead of setting it up as well as re-auth. Is this something Go does not support?
> 
sql.DB is a pool of connection, not a single one. If we try to send multiple requests at the same time the sdk will try to open multiple connections (and then use the cache). I think it's a possible real-world scenario (eg: some BI tools use multiple connections and send multiple requests at the same time to load widgets faster). 

> > Caching in a local file is less performant and having the token cached in memory simplifies the implementation (no security issue of having a token in a file, no having to deal with the lifecycle of the token when it is expired as it's handled by the lib and no potential permission issue when trying to write a file in the disk of the host).
> 
> I see your point, but these are the issues that we've already faced with the Python SDK. Reading from file is slow, but the price is paid only on startup or when token is changed, otherwise it's cached in memory. In there token caching operates on a best effort scenario, if the disk is read-only it does not fail and instead caches in memory. The bottom line is, I'd rather have a consistent approach to caching across different SDKs than each one doing its own thing.

So if we have a BI tool trying to load 10 widgets , if the token is expired or if the token is not available yet we may have 10 threads (or multiple instances of an app) trying to create/modify the same file at the exact the same time - I think it's a bit risky... Plus, the server might return a 401 for a token that is supposedly still valid (based on the expiry time) - and in that scenario we'd have to replace it from the 2 caches (from the memory and from the file) - I think this adds a lot of complexity for little benefit. The JDBC driver doesn't have a cache in a local file either (which is compliant with Postgres)

File: auth.go [hunk 4]
```diff
@@ -13,27 +16,54 @@ type AuthenticationResponse struct {
...
 
-// Authenticate sends an authentication request, and returns a newly constructed client object
+var cache = ttlcache.New[string, string]()
+
+// Authenticate sends an authentication request when a token is not available in the cache, and returns a newly constructed client object
 func Authenticate(username, password, apiEndpoint string) (*Client, error) {
 	infolog.Printf("Start authentication into '%s' using '%s'", apiEndpoint, LoginUrl)
+	userAgent := ConstructUserAgentString()
+	cachedToken := getCachedAccessToken(username, apiEndpoint)
+	if len(cachedToken) > 0 {
```
Comment (ptiurin) [inline auth.go:26]: Nit: would it be clearer to return `nil` from `getCachedAccessToken` if not found? Than checking the length each time.

File: auth.go [hunk 5]
```diff
@@ -13,27 +16,54 @@ type AuthenticationResponse struct {
...
+		infolog.Printf("Returning auth token from cache")
```
Comment (ptiurin) [inline auth.go:27]: maybe this is better at a debug level? I feel like this is too much info for info level (pardon my info).

**Review by ptiurin — APPROVED** [review]
A couple of minor comments, apart from that LGTM.

File: auth.go (hunk 5 above)
Comment (aymeric-dispa) [reply auth.go:27]: Yep I will change that ! I don't think it's even needed

File: auth.go (hunk 4 above)
Comment (ptiurin) [reply auth.go:26]: I see, then possibly `String, error` as a return type would be a more go-like approach?

Comment (aymeric-dispa) [reply auth.go:26]: But not having a token is acceptable and is not an error. Starting to use an error when a string is empty would mean having to treat errors differently when the type is a string. It's perfectly ok to return empty, and checking for emptiness is fine too (that's what they do in the [standard libs](http://golang.org/src/pkg/strconv/atoi.go))
Since we already return "" more than 20 times in the sdk, I think it is best to follow the same coding conventions
